src/scitokens/utils/keycache.py

```diff
@@ -233,26 +234,49 @@ def _get_issuer_publickey(issuer, key_id=None, insecure=False):
         headers={'User-Agent' : 'SciTokens/{}'.format(PKG_VERSION)}
 
         # Go to the issuer's website, and download the OAuth well known bits
-        # https://tools.ietf.org/html/draft-ietf-oauth-discovery-07
-        well_known_uri = ".well-known/openid-configuration"
-        if not issuer.endswith("/"):
-            issuer = issuer + "/"
-        parsed_url = urlparse.urlparse(issuer)
-        updated_url = urlparse.urljoin(parsed_url.path, well_known_uri)
-        parsed_url_list = list(parsed_url)
-        parsed_url_list[2] = updated_url
-        meta_uri = urlparse.urlunparse(parsed_url_list)
-
-        # Make sure the protocol is https
-        if not insecure:
-            parsed_url = urlparse.urlparse(meta_uri)
-            if parsed_url.scheme != "https":
-                raise NonHTTPSIssuer("Issuer is not over HTTPS.  RFC requires it to be over HTTPS")
-        response = request.urlopen(request.Request(meta_uri, headers=headers))
-        data = json.loads(response.read().decode('utf-8'))
+        # https://tools.ietf.org/html/rfc8414
+        # We first try the RFC's metadata location, then try the openid defined
+        well_known_uri = [".well-known/oauth-authorization-server", ".well-known/openid-configuration"]
+        last_exception = MissingIssuerException("Unable to retrieve public key from issuer")
+        jwks_uri = None
+        for uri in well_known_uri:
```

Comment:
This code block is getting long.  Since there is less commonality between OAuth2 and OIDC than thought, maybe make these into two separate subroutines?